bin/_docker.sh

```diff
@@ -50,6 +51,8 @@ docker_build() {
         output="/dev/stderr"
     fi
 
+    validate_tags "$file"
```

Comment:
I like that this makes it easy to validate all docker builds in one place, but I think there's an ordering issue. For instance, if I modify Gopkg.lock and then run `bin/docker-build-proxy-init latest`, that script will first build a new deps image to account for the Gopkg.lock change, even though that image isn't referenced in `proxy-init/Dockerfile`. And then validation happens only after the new, unused deps image has been built. The output of the script looks like this:

```
Successfully tagged gcr.io/runconduit/go-deps:d93a00b2
Tag in proxy-init/Dockerfile does not match source tree:
gcr.io/runconduit/go-deps:41719552 (proxy-init/Dockerfile)
gcr.io/runconduit/go-deps:d93a00b2 (source)
```

I think the validation should happen before we attempt to fetch or build the base image, in which case, the call to `validate_tags` should be happening inside the `docker-build-proxy-init` script itself. That way we're not wasting any time retrieving or building base images that aren't referenced by the dockerfile that we're trying to build.

So I'd revert the changes to this file, and update `docker-build-cli-bin`, `docker-build-controller`, `docker-build-proxy-init`, `docker-build-web`, and `bin/docker-build-proxy` accordingly.

Reply:
I thought about this approach, but settled on the current implementation for a couple reasons:

1. If someone modifies `Gopkg.lock`, they will need to rebuild the deps image. Whether they get a validation error before or after the deps build will not affect the time taken. If we fail the build immediately, they'll be instructed to set an image tag in a Dockerfile for an image that does not yet exist, which may be a bit confusing. I think this order of operations is more instructive to the developer:
  a. Modify `Gopkg.lock`
  b. Build deps image, see deps image, get validation error
  c. Update relevant Dockerfile to reflect new deps image that just got built on my system.

2. Putting validation at a higher level in the `docker-build-*` scripts duplicates implementation, and also runs the risk of someone forgetting to add validation into a new `docker-build-*` script.

3. Keeping the validation in `bin/_tag.sh` keeps all tag/sha-related logic in one place.

Let me know what you think (and if I'm missing something).

Reply:
Hmm, yeah, all good points. I think either approach works. One comment regarding:

> Keeping the validation in `bin/_tag.sh` keeps all tag/sha-related logic in one place.

That file is already sourced in all of the `docker-build-*` scripts, so you could leave the validation code that you wrote in that file, and I agree that that's where it belongs. My recommendation was to remove the calls to validate from `bin/_docker.sh`, and put them in the `docker-build-*` files instead. It's a bit unexpected (to me) that, as a result of your change, the `bin/_docker.sh` script now has a dependency on the `bin/_tag.sh` script. It feels like building and tagging should be handled independently, by the `docker-build-*` scripts themselves.

Reply:
@klingerf per offline convo, have moved validation checks up into the `docker-build-*` files, and removed the `_docker.sh` -> `_tag.sh` dependency. i've added a note in `bin/_tag.sh` providing context on which scripts should do validation checks. we can move that information into a BUILD.md file later on.